text/0626-invoke-helper.md

```diff
@@ -142,124 +118,252 @@ function myProgram(data = []) {
+    import Component from '@ember/component';
+    import { fetch, cancelFetch } from 'fetch-with-cancel';
+
+    export default class FetchData extends Component {
+      @tracked data;
+
+      didReceiveAttrs() {
+        // Cancel the previous fetch if it's still running
+        cancelFetch(this.promise);
+
+        // create a new fetch request, and set the data property to the response
+        this.promise = fetch(this.url)
+          .then((response) => response.json());
+          .then((data) => this.data = data);
+      }
+
+      willDestroy() {
+        cancelFetch(this.promise);
+      }
+    }
+    ```
+    ```hbs
+    <!-- app/components/fetch-data.hbs -->
+    {{yield this.data}}
+    ```
+    ```hbs
+    <!-- app/components/search.hbs -->
+    <DocTitle @title='Search Result for "{{@query}}"'/>
+
+    <FetchData @url="www.example.com/search?query={{@query}}" as |data|>
+      ...
+    </FetchData>
+    ```
+
+   This structure is also not ideal because the components aren't being used for
+   templating, they're just being used for logic effectively. So, there's no
+   reason to add the overhead of a component here.
+
+3. We could use other template constructs, such as helpers or modifiers. Both
+   helpers and modifiers have lifecycles, like components, and can be used to
+   contain functionality. Modifiers aren't really a good choice here though,
+   because it would require us to add an element that we don't need. So, helpers
+   are the better option.
+
+   Helpers work, but like components they require us to move some of our logic
+   into the template, even if that isn't really necessary:
+
+    ```js
+    // app/helpers/doc-title.js
+    import { helper } from '@ember/component/helper';
+
+    export default helper(([title]) => {
+      document.title = title;
+    });
+    ```
+    ```js
+    // app/helpers/fetch-data.js
+    import Helper from '@ember/component/helper';
+    import { fetch, cancelFetch } from 'fetch-with-cancel';
+
+    export default class FetchData extends Helper {
+      @tracked data;
+
+      compute([url]) {
+        if (this._url !== url) {
+          this.url = url;
+
+          // Cancel the previous fetch if it's still running
+          cancelFetch(this.promise);
+
+          // create a new fetch request, and set the data property to the response
+          this.promise = fetch(url)
+            .then((response) => response.json());
+            .then((data) => this.data = data);
+        }
+
+        return this.data;
+      }
+
+      willDestroy() {
+        cancelFetch(this.promise);
+      }
+    }
+    ```
+    ```hbs
+    <!-- app/components/search.hbs -->
+    {{doc-title 'Search Result for "{{@query}}"'}}
+
+    {{#let (fetch-data "www.example.com/search?query={{@query}}") as |data|}}
+      ...
+    {{/let}}
+    ```
+
+Out of these options, helpers are the closest to what we want - they are low
+overhead, they produce computed values directly without a template, and
```

Comment:
FWIW, I strongly dislike the "overhead" argument. I just don't see any rationale for how helpers are inherently lower overhead than a component in modern Ember (e.g. the Octane mental model). Both helpers and components (ala `@glimmer/component`) are _effectively_ the cost of doing `new SomeThing()` where `SomeThing` has a very shallow inheritance structure and very little code in `constructor`.